New Entity: target_weapons.

lumayaa opened the pull request:
The activator can be given these weapons: stun_baton, melee, single | duals | staff, bryar_pistol, blaster, disruptor, bowcaster, repeater, demp2, flechette, rocket_launcher, thermal ,tripmine, detpack, concussion.

Example: /place target_weapons * targetname,asd,weapons,single.detpack5.rocket_launcher.thermal999.disruptor0

The example above would do the following to the activator:
give him single saber, and locks him into it until unlocked or death
give him 5 detpacks
give him rocket_launcher with INFINITE Ammo
give him 999 thermals
remove disruptor rifle if he has any

Spawnflags:

- "1", "Don't lock the saber, meaning the given saber can be switched from duals to staff for instance."
-  "2", "Just allow the weapons but don't give them."
-  "4", "Add the ammo instead of setting it."

Keys:

- weapons", "Given Weapons. Example: weapons,detpack5.concussion5000.single - this would give the activator detpack with 5 ammo concussion with 5000 ammo. if no number is given then its infinite. staff|duals|single take no ammo. if 0 is given then it removes the weapon,
- targetname", "make the trigger target this value for the entity to be used

lumayaa committed:
Init.

lumayaa committed:
Added 2 more spawnflags.

slash-fr commented:
I noticed a few issues:

1. Spawnflag 1 seems to do the opposite of what it says. (1 actually locks the saber?)

2. `target_weapon` seems to make the `/saber` command unusable, even after respawn, in some circumstances.  
   "Unusable" means that the userinfo does not update when using `/saber`. But changing the saber from the UI works.  
   I’m not completely sure what conditions trigger the bug. I think it involves adding a saber to the `weapon` key, and avoiding spawnflag 4?

3. The description of the `weapons` key says: `"if no number is given then its infinite."`  
   `999` would be more accurate than `infinite`.  
   With a weapon such as the concussion rifle, 999 ammo can be used up in less than 20 seconds (so, not quite infinite ^^).
   If you want to increase this number, the maximum that seems to work is `SHRT_MAX` (32767).  
   
   Which brings me to the next point:

4. If a builder uses an ammo number larger than `SHRT_MAX` (32767), or if the ammo amount goes over 32767 because of spawnflag 4 (*"Add the ammo instead of setting it"*), the weapon becomes unusable (as if it had no ammo).

slash-fr commented on game/g_target.c:
```diff
@@ -235,31 +263,275 @@ void SP_target_delay( gentity_t *ent ) {
+            }
+
+            if (weaponID != WP_SABER && weaponID != WP_BRYAR_PISTOL)
+                activator->client->ps.ammo[weaponData[weaponID].ammoIndex] = addAmmo ? activator->client->ps.ammo[weaponData[weaponID].ammoIndex] + ammoAmount : ammoAmount;
```
I would recommend capping the ammo to `SHRT_MAX` (to ensure the weapon still works):
```c
if (activator->client->ps.ammo[weaponData[weaponID].ammoIndex] > SHRT_MAX)
{
    activator->client->ps.ammo[weaponData[weaponID].ammoIndex] = SHRT_MAX;
}
```

lumayaa committed:
Fixes.

Infinite Ammo amount changed from 999 to SHRT_MAX.
Capped max ammo to SHRT_MAX.
Fixed spawnflags 1 doing the opposite of what it says it does.

lumayaa commented:
> I noticed a few issues:
> 
> 1. Spawnflag 1 seems to do the opposite of what it says. (1 actually locks the saber?)
> 2. `target_weapon` seems to make the `/saber` command unusable, even after respawn, in some circumstances.
>    "Unusable" means that the userinfo does not update when using `/saber`. But changing the saber from the UI works.
>    I’m not completely sure what conditions trigger the bug. I think it involves adding a saber to the `weapon` key, and avoiding spawnflag 4?
> 3. The description of the `weapons` key says: `"if no number is given then its infinite."`
>    `999` would be more accurate than `infinite`.
>    With a weapon such as the concussion rifle, 999 ammo can be used up in less than 20 seconds (so, not quite infinite ^^).
>    If you want to increase this number, the maximum that seems to work is `SHRT_MAX` (32767).
>    Which brings me to the next point:
> 4. If a builder uses an ammo number larger than `SHRT_MAX` (32767), or if the ammo amount goes over 32767 because of spawnflag 4 (_"Add the ammo instead of setting it"_), the weapon becomes unusable (as if it had no ammo).

Regarding that 2nd Bug. Probably not looking into that rn. Rest should be fixed.

lumayaa committed:
Merge branch 'develop' into SetSaberEntity
